samples: Merge Azure IoT Hub and Azure FOTA samples (#11501)
nrfconnect/sdk-nrf: samples/net/azure_iot_hub/README.rst

For the sample to work as intended, you must set up and configure an Azure IoT Hub instance.
See :ref:`configure_options_azure_iot` for information on the configuration options that you can use to create an Azure IoT Hub instance.
Also, for a successful TLS connection to the Azure IoT Hub instance, the device needs to have certificates provisioned.

> If you want to test FOTA, ensure that also the required credentials for the file server are provisioned and that the :kconfig:option:`CONFIG_AZURE_FOTA_SEC_TAG` Kconfig option is set accordingly.

Reviewer: ```suggestion
If you want to test FOTA, ensure that also the required credentials for the file server are provisioned and the :kconfig:option:`CONFIG_AZURE_FOTA_SEC_TAG` Kconfig option is set accordingly.
```

* :kconfig:option:`CONFIG_AZURE_IOT_HUB_DPS_REG_ID` - Sets the Azure IoT Hub DPS registration ID. It can be provided at run time. By default, the sample uses the device ID as the registration ID and sets it at run time.
* :kconfig:option:`CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE` - Sets the DPS ID scope of the Azure IoT Hub. This can be provided at run time.

For FOTA, check and configure the following library Kconfig options:

* :kconfig:option:`CONFIG_AZURE_FOTA_APP_VERSION` - Defines the application version string. Indicates the current firmware version on the development kit.
* :kconfig:option:`CONFIG_AZURE_FOTA_APP_VERSION_AUTO` - Automatically generates the application version. If enabled, :kconfig:option:`CONFIG_AZURE_FOTA_APP_VERSION` is ignored.
* :kconfig:option:`CONFIG_AZURE_FOTA_TLS` - Enables HTTPS for downloads. By default, TLS is enabled and currently, the transport protocol must be configured at compile time.

> * :kconfig:option:`CONFIG_AZURE_FOTA_SEC_TAG` - Sets the security tag for TLS credentials when using HTTPS as the transport layer. See :ref:`certificates` for more details.

Reviewer 2: yes, it seems to reference the old README file of the Azure Fota sample.. I can move the certification text to this sample instead.

> * :kconfig:option:`CONFIG_AZURE_FOTA_SEC_TAG` - Sets the security tag for TLS credentials when using HTTPS as the transport layer. See :ref:`certificates` for more details.

Reviewer 2: or refer to the `prereq_connect_to_azure_iot_hub` instead, is being referred to in the credential reference anyway.

#. Observe that the event messages from the device are displayed in the terminal within the specified telemetry interval.

To test FOTA update:

1. Using the `Azure IoT Explorer`_ or `Azure Portal`_, navigate to the view where you can update the device twin.
#. Create a `firmware` JSON object inside the *desired* object of the device twin document, containing information about the binary to download:

   .. parsed-literal::
      :class: highlight

	{
	    "firmware": {
	        "fwVersion": "v0.0.2-dev",
	        "jobId": "ca186d4b-4171-4209-a49d-700c35567d1d",
	        "fwLocation": {
	            "host": "my-fota-server.comt",
	            "path": "firmware/my-app-update.bin"
	        },
	        "fwFragmentSize": 1800
	    }
	}

#. Apply the changes to the device twin by clicking :guilabel:`Save`. Updating the device twin causes a message to be sent to the device, containing the updates to the *desired* object.
#. In the terminal emulator, observe that the new firmware image is downloaded and installed as shown in :ref:`sampoutput_azure_iot`.
#. When the development kit reboots, observe that the sample displays the logs from the new application.

> #. In `Azure IoT Explorer`_ or Azure Portal, confirm that the device twin contains the updated application version in the ``reported.firmware.fwVersion`` field.

Reviewer: ```suggestion
#. Confirm that the device twin contains the updated application version in the ``reported.firmware.fwVersion`` field in `Azure IoT Explorer`_ or `Azure Portal`_.
```

#. In the `Azure IoT Explorer`_, navigate to the :guilabel:`Telemetry` tab and click :guilabel:`start`.
#. Observe that the event messages from the device are displayed in the terminal within the specified telemetry interval.

To test FOTA update:

1. Navigate to the view where you can update the device twin using the `Azure IoT Explorer`_ or `Azure Portal`_.
#. Create a ``firmware`` JSON object inside the *desired* object of the device twin document, containing information about the binary to download:

   .. parsed-literal::
      :class: highlight

	{
	    "firmware": {
	        "fwVersion": "v0.0.2-dev",
	        "jobId": "ca186d4b-4171-4209-a49d-700c35567d1d",
	        "fwLocation": {

> 	            "host": "my-fota-server.comt",

Reviewer 3: ```suggestion
	            "host": "example.com",
```

Use generic example domain.

	            "path": "firmware/my-app-update.bin"

> 	        },

Reviewer 3: **question:** can't we use a fully qualified URL by now? AFAIR the download client is now able to use that?

	        "fwFragmentSize": 1800
	    }
	}

#. Click :guilabel:`Save` to apply the changes to the device twin.
   Updating the device twin causes a message to be sent to the device, containing the updates to the *desired* object.
#. In the terminal emulator, observe that the new firmware image is downloaded and installed as shown in :ref:`sampoutput_azure_iot`.
#. Observe that the sample displays the logs from the new application when the development kit reboots.
#. Confirm that the device twin contains the updated application version in the ``reported.firmware.fwVersion`` field in `Azure IoT Explorer`_ or `Azure Portal`_.

.. _sampoutput_azure_iot:

Sample output
=============

> When the sample runs, the device boots, and the sample displays the following output in the terminal over UART on an nRF7002 DK:

Reviewer 3: *question:* Why is `nRF7002` relevant?

	<inf> azure_iot_hub_sample: Event was successfully sent
	<inf> azure_iot_hub_sample: Next event will be sent in 60 seconds

If a new FOTA update is initiated, the console output is like this:

.. code-block:: console

	<inf> azure_iot_hub_sample: AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS, ID: 140

> 	<inf> azure_fota: Attempting to download firmware (version 'v0.0.2-dev') from my-fota-server.com/firmware/app_update.bin

Reviewer 3: ```suggestion
	<inf> azure_fota: Attempting to download firmware (version 'v0.0.2-dev') from example.com/firmware/app_update.bin
```

	<inf> download_client: Downloading: firmware/app_update.bin [0]
	<inf> azure_iot_hub_sample: AZURE_IOT_HUB_EVT_FOTA_START
	<inf> azure_iot_hub_sample: AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED
	<inf> download_client: Setting up TLS credentials, sec tag count 1

> 	<inf> download_client: Connecting to my-fota-server.com

Reviewer 3: ```suggestion
	<inf> download_client: Connecting to example.com
```
